DM-2528 - LWM2M X.509 documentation (#1420)
SoftwareAG/c8y-docs: content/protocol-integration/lwm2m-bundle/registering-lwm2m-devices.md

When you upload the CSV file, the dialog will display a confirmation message that tells you where to find the result. The bulk device registration operation will be displayed in the [LWM2M connector device](#connector-device) object created for the tenant.

> #### Device settings {#lwm2m-device-registration-device}

Author: "device-settings" seems too generic. I was afraid that it may be used in other parts of documentation. I'm not sure how the links/hooks are generated so I went with something safe/unique.

openssl pkcs8 -topk8 -inform PEM -outform PEM -in myDevice.key -out myDevice.key.pem -nocrypt
# optionally export to DER format if your device needs it
openssl pkcs8 -topk8 -inform PEM -outform DER -in myDevice.key -out myDevice.der -nocrypt
```

##### Trusting CA in {{< product-c8y-iot >}}
Before devices are able to connect to the platform, CA that issued device certificates must be added to trusted certificates. See [Managing trusted certificates](/users-guide/device-management/#trusted-certificates) on how to add and trust CA certificate.

##### All security fields details
The fields below must be contained to configure security modes for bootstrap and regular server connection:
<table>
<col>
<col style="width:30%">
<col style="width:11%">
<col style="width:30%">
<col>
<thead>
<tr>
<th style="text-align:left">Field</th>
<th style="text-align:left">CSV field name</th>
<th style="text-align: left">Type</th>
<th style="text-align:left">Description</th>
<th style="text-align: left">Mandatory</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Server security mode</td>
<td style="text-align:left">securityMode</td>
<td style="text-align: left">String</td>

> <td style="text-align:left">Determines the type of connection used by the LWM2M device when it connects to {{< product-c8y-iot >}} LWM2M server. Possible values are: "NO_SEC", "PSK" and "X509".</td>

Reviewer: ```suggestion
<td style="text-align:left">Determines the type of connection used by the LWM2M device when it connects to the {{< product-c8y-iot >}} LWM2M server. Possible values are: "NO_SEC", "PSK" and "X509".</td>
```

<td style="text-align: left">Mandatory</td>
</tr>
<tr>
<td style="text-align:left">Bootstrap security mode</td>
<td style="text-align:left">bootstrapSecurityMode</td>
<td style="text-align: left">String</td>

> <td style="text-align:left">Determines the type of connection used by the LWM2M device when it connects to {{< product-c8y-iot >}} LWM2M bootstrap server. Possible values are: "NO_SEC", "PSK", "PSK_GENERATED" and "X509".</td>

Reviewer: ```suggestion
<td style="text-align:left">Determines the type of connection used by the LWM2M device when it connects to the {{< product-c8y-iot >}} LWM2M bootstrap server. Possible values are: "NO_SEC", "PSK", "PSK_GENERATED" and "X509".</td>
```

<td style="text-align: left">The hex-encoded pre-shared key used by the device for server connections in PSK mode.</td>
<td style="text-align: left; height: 26px;">Mandatory for PSK security mode.</td>
</tr>
<tr>
<td style="text-align: left">Bootstrap PSK ID</td>
<td style="text-align: left">bootstrap psk_id</td>
<td style="text-align: left">String</td>
<td style="text-align: left">The ID used by the device for bootstrap connections in PSK mode. The bootstrap PSK ID has be to be <b>unique</b> across all tenants. Registering a device using an bootstrap PSK ID already used will result in an error.</td>
<td style="text-align: left">Mandatory for PSK security mode.</td>
</tr>
<tr>
<td style="text-align: left">Bootstrap PSK key</td>
<td style="text-align: left">bootstrap psk_key</td>
<td style="text-align: left">String</td>
<td style="text-align: left">The hex-encoded key used by the device for bootstrap connections in PSK mode.</td>
<td style="text-align: left">Mandatory for PSK security mode.</td>
</tr>
<tr>
<td style="text-align: left">X.509 certificate</td>
<td style="text-align: left">x509ClientCertificate</td>
<td style="text-align: left">String</td>

> <td style="text-align: left">X.509 device certificate (in PEM format) written to device during bootstrap phase. Empty value means that it won't be written at all.</td>

Reviewer: ```suggestion
<td style="text-align: left">X.509 device certificate (in PEM format) written to the device during bootstrap phase. An empty value means that it isn't written at all.</td>
```

<td style="text-align: left">Optional</td>
</tr>
<tr>
<td style="text-align: left">X.509 private key</td>
<td style="text-align: left">x509ClientPrivateKey</td>
<td style="text-align: left">String</td>

> <td style="text-align: left">X.509 device private key (in PEM format) written to device during bootstrap phase. Empty value means that it won't be written at all.</td>

Reviewer: ```suggestion
<td style="text-align: left">X.509 device private key (in PEM format) written to the device during bootstrap phase. An empty value means that it isn't written at all.</td>
```

<td style="text-align: left">Optional</td>
</tr>
<tr>
<td style="text-align: left">X.509 certificate usage</td>
<td style="text-align: left">x509CertificateUsage</td>
<td style="text-align: left">String</td>

> <td style="text-align: left">LWM2M Certificate Usage written to device during bootstrap phase. Empty value means that it won't be written at all. One of: <ul><li>CA_CONSTRAINT</li> <li>SERVICE_CERTIFICATE_CONSTRAINT</li> <li>TRUST_ANCHOR_ASSERTION</li> <li>DOMAIN_ISSUER_CERTIFICATE</li></ul></td>

Reviewer: ```suggestion
<td style="text-align: left">LWM2M Certificate usage written to the device during bootstrap phase. An empty value means that it isn't written at all. One of: <ul><li>CA_CONSTRAINT</li> <li>SERVICE_CERTIFICATE_CONSTRAINT</li> <li>TRUST_ANCHOR_ASSERTION</li> <li>DOMAIN_ISSUER_CERTIFICATE</li></ul></td>
```

<td style="text-align: left">Optional</td>
</tr>
<tr>
<td style="text-align: left">Server certificate</td>
<td style="text-align: left">serverPublicKey</td>
<td style="text-align: left">String</td>

> <td style="text-align: left">Name of the server certificate to write to device during bootstrap phase. Server certificates are preconfigured in "management" tenant. Empty value means that it won't be written at all.</td>

Reviewer: ```suggestion
<td style="text-align: left">Name of the server certificate written to the device during bootstrap phase. Server certificates are preconfigured in the {{< management-tenant >}}. An empty value means that it isn't written at all.</td>
```

</table>

{{< c8y-admon-info >}}
After creation, you can view and change device parameters in the **LWM2M configuration** tab in the **Device details** page, see [LWM2M configuration](#lwm2m-configuration).
{{< /c8y-admon-info >}}

> #### Advanced settings {#lwm2m-device-registration-advanced}

Reviewer 2: In advanced settings table, these ones needs to be updated:

- "Use common timestamp resources 5518 and 6050 or object specific timestamp for object 6 reported by the device if available" -> Use source timestamp
- "Device protocol serialization format" -> Serialization format

Author: done
